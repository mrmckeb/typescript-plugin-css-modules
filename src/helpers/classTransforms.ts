import camelCase from 'lodash.camelcase';
import { ClassnameTransformOptions } from '../options';

// The below is based on the CSS Modules implementation found here:
// https://github.com/webpack-contrib/css-loader

const dashCase = (classname: string): string =>
  classname.replace(/-+(\w)/g, (_match: string, firstLetter: string) =>
    firstLetter.toUpperCase(),
  );

export const transformClasses =
  (camelCaseOption?: ClassnameTransformOptions) =>
  (classname: string): string[] => {
    const entries: string[] = [];

    switch (camelCaseOption) {
      case 'camelCase': {
        entries.push(classname);
        const targetClassName = camelCase(classname);
        if (targetClassName !== classname) {
          entries.push(targetClassName);
        }
        break;
      }
      case 'camelCaseOnly':
        entries.push(camelCase(classname));
        break;
      case 'dashes': {
        entries.push(classname);
        const targetClassName = dashCase(classname);
        if (targetClassName !== classname) {
          entries.push(targetClassName);
        }
        break;
      }
      case 'dashesOnly':
        entries.push(dashCase(classname));
        break;
      case 'asIs':
      default:
        entries.push(classname);
        break;
    }

    return entries;
  };
