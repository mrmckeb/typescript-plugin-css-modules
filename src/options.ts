import { Options as SassOptions } from 'sass';
import tsModule from 'typescript/lib/tsserverlibrary';
import { DotenvConfigOptions } from 'dotenv/types';
import { CSSExports } from 'icss-utils';
import stylus from 'stylus';
import { Logger } from './helpers/logger';

// NOTE: Stylus doesn't directly export RenderOptions.
type StylusRenderOptions = Parameters<typeof stylus>[1];

export interface PostcssOptions {
  excludePlugins?: string[];
  useConfig?: boolean;
}

export interface RendererOptions {
  less?: Partial<Less.Options>;
  sass?: Partial<SassOptions<'sync'>>;
  stylus?: Partial<StylusRenderOptions>;
}

export interface Options {
  classnameTransform?: ClassnameTransformOptions;
  customMatcher?: string;
  customRenderer?: string;
  customTemplate?: string;
  dotenvOptions?: DotenvConfigOptions;
  namedExports?: boolean;
  postcssOptions?: PostcssOptions;
  /** @deprecated To align with other projects. */
  postCssOptions?: PostcssOptions;
  rendererOptions?: RendererOptions;
}

export type ClassnameTransformOptions =
  | 'asIs'
  | 'camelCase'
  | 'camelCaseOnly'
  | 'dashes'
  | 'dashesOnly';

export interface CustomRendererOptions {
  fileName: string;
  logger: Logger;
  compilerOptions: tsModule.CompilerOptions;
}

export type CustomRenderer = (
  css: string,
  options: CustomRendererOptions,
) => string;

export interface CustomTemplateOptions {
  classes: CSSExports;
  fileName: string;
  logger: Logger;
}

export type CustomTemplate = (
  dts: string,
  options: CustomTemplateOptions,
) => string;
