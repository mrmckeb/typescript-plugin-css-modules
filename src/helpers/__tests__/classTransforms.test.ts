import { transformClasses } from '../classTransforms';
import { ClassnameTransformOptions } from '../../options';

describe('utils / classTransforms', () => {
  const classnames = [
    'class-name-a',
    'classNameB',
    'class-Name-C',
    '__class_nAmeD--',
  ];
  const tests: ClassnameTransformOptions[] = [
    'camelCase',
    'camelCaseOnly',
    'dashes',
    'dashesOnly',
    'asIs',
  ];

  it('should not transform classes when no option is set', () => {
    const transformer = transformClasses();
    const transformedClasses = classnames.map(transformer);
    expect(transformedClasses).toMatchSnapshot();
  });

  tests.forEach((option) => {
    it(`should transform classes correctly when \`classnameTransform\` set to \`${option}\``, () => {
      const transformer = transformClasses(option);
      const transformedClasses = classnames.map(transformer);
      expect(transformedClasses).toMatchSnapshot();
    });
  });
});
