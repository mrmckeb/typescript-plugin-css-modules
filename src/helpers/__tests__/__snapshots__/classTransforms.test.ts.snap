// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helpers / classTransforms should not transform classes when no option is set 1`] = `
[
  [
    "class-name-a",
  ],
  [
    "classNameB",
  ],
  [
    "class-Name-C",
  ],
  [
    "__class_nAmeD--",
  ],
]
`;

exports[`helpers / classTransforms should transform classes correctly when \`classnameTransform\` set to \`asIs\` 1`] = `
[
  [
    "class-name-a",
  ],
  [
    "classNameB",
  ],
  [
    "class-Name-C",
  ],
  [
    "__class_nAmeD--",
  ],
]
`;

exports[`helpers / classTransforms should transform classes correctly when \`classnameTransform\` set to \`camelCase\` 1`] = `
[
  [
    "class-name-a",
    "classNameA",
  ],
  [
    "classNameB",
  ],
  [
    "class-Name-C",
    "classNameC",
  ],
  [
    "__class_nAmeD--",
    "classNAmeD",
  ],
]
`;

exports[`helpers / classTransforms should transform classes correctly when \`classnameTransform\` set to \`camelCaseOnly\` 1`] = `
[
  [
    "classNameA",
  ],
  [
    "classNameB",
  ],
  [
    "classNameC",
  ],
  [
    "classNAmeD",
  ],
]
`;

exports[`helpers / classTransforms should transform classes correctly when \`classnameTransform\` set to \`dashes\` 1`] = `
[
  [
    "class-name-a",
    "classNameA",
  ],
  [
    "classNameB",
  ],
  [
    "class-Name-C",
    "classNameC",
  ],
  [
    "__class_nAmeD--",
  ],
]
`;

exports[`helpers / classTransforms should transform classes correctly when \`classnameTransform\` set to \`dashesOnly\` 1`] = `
[
  [
    "classNameA",
  ],
  [
    "classNameB",
  ],
  [
    "classNameC",
  ],
  [
    "__class_nAmeD--",
  ],
]
`;
